### DISCUSSION TITLE
How to log 500 error/exception from other services?

### DISCUSSION THREAD

**Discussion by tungphuong** [ThreeMammals/Ocelot#1819] [Q&A] [answered]
Hello,

I have just used it. It works well but I have one concern that how to log exception (Http code 500) from downstream services. Currently, if I switch log level to "error", I cannot receive anything. However, if log level is "trace", there is some below info
```
trce: Ocelot.Logging.OcelotDiagnosticListener[0]
      => RequestId:0HLFHAF0APTOQ:00000002 RequestPath:/api/r/xxx/countries => RequestId:0HLFHAF0APTOQ:00000002 RequestPath:/api/r/xxx/countries => RequestId:0HLFHAF0APTOQ:00000002 RequestPath:/api/r/xxx/countries
      requestId: 0HLFHAF0APTOQ:00000002, previousRequestId: no previous request id, message: MiddlewareFinished: Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware; 500
: Ocelot.Logging.OcelotDiagnosticListener[0]
      requestId: 0HLFHAF0APTOQ:00000002, previousRequestId: no previous request id, message: MiddlewareFinished: TransitionToOcelotMiddleware; 500
```
Please let me know if there is any missing.

Thanks,

**Comment by TomPallister** [comment]
@tungphuong I will test this and get back to you ASAP! If you want to submit a PR that adds the logging let me know.

**Comment by TomPallister** [comment]
from other issue

```
Hi,

Currently, if downstream service throw exception with httpcode 500, but Ocelot just logs "warning" or "debug". It will not clear and difficult to trace/monitor on production. Because we cannot set log level is debug/info for production.

I suggest we should check status code. If it is 500, we wil log like error.\

Cheers,

Phuong Le
```

**Comment by tungphuong** [comment]
Yes, 2 items are the same. Ok, I will submit one PR :)

**Reply by raman-m** [reply]
Did you submit a PR?

**Reply by raman-m** [reply]
5 years have been passed since submission!... 😸

**Comment by raman-m** [comment]
> @tungphuong [on Jul 24, 2018](#discussion-5906981)

Are you still with Ocelot?
And I don't see Ocelot being forked in [your repos](https://github.com/tungphuong?tab=repositories)...

**Comment by ks1990cn** [comment]
Interesting, can I replicate on AWS EC2 with cloudwatch? Such that nothing is found on cloudwatch when loglevel set as "Error"? 

Can I try replicating issue on this case?

**Reply by raman-m** [reply]
Tanmay, what's the goal of this discussion?
I've converted old issue to this discussion a month ago having a hope the author will reply with a feedback... Finally, he is silent...
I don't see any sense...
Do you see?

**Comment by raman-m** [comment, marked as answer]
@tungphuong [on Jul 24, 2018](https://github.com/ThreeMammals/Ocelot/discussions/1819#discussion-5906981):
> I have one concern that how to log exception (Http code 500) from downstream services.

Hey Phuong Le!
Seems you are no longer with Ocelot... This is last try to ping you...

What details of 500 error would you like to log by Ocelot? 
- Status code and the path
- Status code, template path, service error/exception details

And, why do you think that Ocelot should log error details of downstream service?
I believe downstream should log service errors.

If you want to transfer root cause of 500 error then you have to control downstream service, develop, setup, and prepare error details model to send all this information in response as JSON for example to allow Ocelot to handle this response, detect the case, and log finally by internal Ocelot logger.

Do you understand that you've described invalid user scenario?

------
> Currently, if I switch log level to "error", I cannot receive anything.

Right, because Ocelot has internal Error Mapper logic, see the feature docs: [Error Status Codes | Design](https://ocelot.readthedocs.io/en/latest/features/errorcodes.html#design)
In case of error 500 on downstream service side Ocelot will recognize it as failed request -> [Server error responses](https://ocelot.readthedocs.io/en/latest/features/errorcodes.html#server-error-responses)

From second side, `Error` level of Ocelot app means you catch internal Ocelot app exceptions. 500 error of downstream seems doesn't produce internal Ocelot error, the error just remapped with 5xx status but nothing logged.
So, the [HttpExceptionToErrorMapper](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Requester/HttpExceptionToErrorMapper.cs) doesn't log anything, simply remapping.
That's why you don't see downstream errors in Ocelot logs!

-----
> However, if log level is "trace", there is some below info

Exactly, because [OcelotDiagnosticListener](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Logging/OcelotDiagnosticListener.cs) logs only if `Trace` level, for debugging purposes. And finally, you realized that you have 500 error case.

----
> `MiddlewareFinished: Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware; 500`

Seems the root cause of downstream service failure is invalid CORS policies of the request! Ups! 💥  Bingo!
**Setup your request with valid CORS headers to be able being accepted by downstream service!**

**Reply by raman-m** [reply]
Once again, again and again...
**Please check direct connection first!** If direct connection to downstream service works well, then start write route JSON, and work on route antifactory Ocelot classes.

I don't understand when people says "The app doesn't work. I have a problem".
Please. don't be a lazy developer to check direct connection first! 😠

**Reply by raman-m** [reply]
> How to log 500 error/exception from other services?

Answering the question...
**First**, you have to send error details in response to the Ocelot's side!
**Second**, write [Delegating Handler](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html) to control the route totally analyzing all properties of the response.
**Third**, your Delegating Handler can be a prototype to process such cases by Ocelot middlewares, and you will be able to create a PR.
